Custom CSV download chooser feature

michaelwood opened the pull request.

Commit:
templates: Rename Insights to GrantVis

Also tidy up the text and alignments

Commit:
views/csv/tags: Backend application changes needed for custom downloads

This adds all the changes needed in the backend to be enable the first
version of the user customisable CSV downloads.
This also maintains the existing download functionality support.

Commit:
templates: Add new customisable CSV download UI

This adds the new modal and widgets for the first version of the
custom csv download chooser.

Commit:
csv_layout: Fix flake8 error

Commit:
templates: dl-field-chooser: Add feature to "grey out" field categories

This updates the disabled state of the field category to show whether
all the fields have been added to the list or not.

Commit:
csv_layout: Fix incorrect field path for retrieved in GN

Commit:
templates: dl-field-chooser: Remove additional "(additional data)"

This is now embedded into the field titles so doesn't need to be added
in the templates.

Commit:
templates: Search: Remove trailing line for eslint

R2ZER0 commented on grantnav/csv_layout.py:
```diff
@@ -30,13 +32,27 @@
+
+    ("Regrant Type", "result.regrantType"),
+
+    ("Funding Type Title", "result.fundingType.0.title"),
```
Did you mean to remove the colon  `Funding Type:Title` -> `Funding Type Title` ? I guess it's not really a category since there is only the one field

R2ZER0 replied:
I see further down that you do split by `:`, hence removing

michaelwood replied:
Yeah being more strict on the ":" being consistent way to represent the object hierarchy. This also helps with flatten/unflatten processing

R2ZER0 commented on grantnav/csv_layout.py:
```diff
@@ -134,3 +153,35 @@
+    """ takes grants_csv and turns it into a dictionary of parent/child
+    fields e.g.
+
+    res["Planned Dates"][
```
is this meant to mean that dict values are arrays?
```suggestion
    res["Planned Dates"] = [
```

michaelwood replied:
yep thanks

R2ZER0 commented on grantnav/csv_layout.py:
```diff
@@ -134,3 +153,35 @@
+    res = {}
+
+    for title, path in grants_csv.items():
+        parsed = title.split(":")
```
Is it possible to have more than one `:` in a title?
```suggestion
        parsed = title.split(":", maxsplit=1)
```

R2ZER0 commented on grantnav/frontend/views.py:
```diff
@@ -167,13 +167,26 @@ def write(self, value):
-    response = StreamingHttpResponse(chain(['\ufeff'], (writer.writerow(row) for row in grants_csv_generator(helpers.clean_for_es(query)))), content_type='text/csv')
+    response = StreamingHttpResponse(
+        chain(
+            ["\ufeff"],
```
Eh?

R2ZER0 replied:
oh it's just a reformat anyway

Commit:
csv_layout: Clarify help text for dict conversion

Co-authored-by: Rikki Guy <rikki.guy@opendataservices.coop>

Commit:
csv_layout: Dict conversion ensure predicable split

We only handle two levels of hierarchy in the CSV customiser so we need to avoid this splitting more than once.

Co-authored-by: Rikki Guy <rikki.guy@opendataservices.coop>

Commit:
templates: Restore the grantnav "button"

This doesn't quite fit but has been requested to be incorporated
into this part of the design.

Commit:
templates: s/360Insights/GrantVis

Also fix typo for testid on insights button.

michaelwood commented:
Branch is now ready for deploy when we have the green light from @mariongalley

Commit:
templates: dl-field-chooser: Fix overlapping scope of CSS

This appears to be causing some issues with other elements in grantnav
and as the scope isn't clear due to it being an included template wrap
the whole thing in a css class and parent the selectors to this new
class.

Commit:
csv_layout: Rename "Publisher:Name" to "Publisher Name"

This makes it consistent with the other top level fields and avoids
introducing an issue delimiting parent:child fields.

Commit:
csv_layout: Remove "Recipient" from the best available geo fields

Commit:
csv_layout: Change "#comment" to "#note"

Commit:
templates: dl-field-chooser: Remove debug output

R2ZER0 commented on grantnav/csv_layout.py:
```diff
@@ -70,45 +86,48 @@
     ("Beneficiary Location:7:Geographic Code Type", "result.beneficiaryLocation.7.geoCodeType"),
     ("From An Open Call?", "result.fromOpenCall"),
-    ("The following fields are not in the 360 Giving Standard and are added by GrantNav.", ""),
+    # ("#comment The following fields are not in the 360 Giving Standard and are added by GrantNav.", ""),
```
Is this one supposed to be `#Note` too?

R2ZER0 replied:
oh it's commented out anyway

R2ZER0 commented on grantnav/frontend/templates/components/results-title.html:
```diff
@@ -4,17 +4,22 @@ <h2 class="results-title">
-    <div class="radio-buttons__group">
+  <div>
+    <div class="radio-buttons export-wrapper" aria-labelledby="view-mode-title" style="align-items: baseline">
+      <h4 class="margin-right:1" id="view-mode-title">View Mode:</h4>
```
What does it mean to have `:1` on the class name here in the HTML?

michaelwood replied:
This is just how the 360 design system utils work - https://github.com/ThreeSixtyGiving/360-ds/blob/master/src/scss/pre/_utils.scss#L81

R2ZER0 commented on grantnav/frontend/user_csv_layout.py:
```diff
@@ -0,0 +1,20 @@
+
+    return views.grants_csv_paged(
+        json.loads(request.POST.get("json_query", {})),
+        grant_csv_titles=[item["column_title"] for item in fields] + grant_csv_titles[-2:],
```
What's special about the last two grant_csv_titles ?

michaelwood replied:
we always want to add the last two which are License and Note fields (non-optional). I've added a comment

Commit:
csv_layout/dl-chooser: Remove "#comment" as this was confusing

Create individual exclusions for the License and Note field from
being included in the optional fields for the custom csv download.

Commit:
templates: download-modal: Bold the amount of grants to download

Commit:
user_csv_layout: Add comment about why the last two fields are added

michaelwood commented:
Couple of extra commits to review @R2ZER0 thanks

R2ZER0 reviewed (approved):
Looks good :+1:
